core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -100,19 +100,20 @@ object ByteBufferMessageSet {
       }
       var lastInnerOffset = -1L
 
-      val messageAndOffsets = if (wrapperMessageAndOffset.message.magic > MagicValue_V0) {
+      val messageAndOffsets = {
         val innerMessageAndOffsets = new ArrayDeque[MessageAndOffset]()
         try {
           while (true)
             innerMessageAndOffsets.add(readMessageFromStream())
         } catch {
           case eofe: EOFException =>
-            compressed.close()
           case ioe: IOException =>
             throw new InvalidMessageException(s"Error while reading message from stream compressed with ${wrapperMessage.compressionCodec}", ioe)
+        } finally {
+          CoreUtils.swallow(compressed.close())
```

Comment:
Maybe we can now clean-up the fields in the Iterator. Things like `compressed` and `inputStream` are only used within the scope of this val. Both of them and any other fields that don't have to be available for the life of the iterator could be moved to this block to make that clear. What do you think?

The rest looks good. I was thinking of how we could test this, but it seems like the kind of thing that we should check via memory leak detection tests.